01-architecture.md

```diff
@@ -0,0 +1,222 @@
+
+Finally, this approach is readily compatible with VMs that rely on internal dispatch (e.g. EVM), but also stretches to accommodates for VMs that perform outer dispatch.
+
+Approach 2 would render the `MethodNum` field on messages obsolete, as external dispatch is no longer necessary, and all call information would be contained in the input parameters.
+
+We _could_ technically preserve the `MethodNum`, and have actors perform internal dispatch based on that _and_ input parameters. But we believe that's unnecessary cruft, and it breaks the actor-orientation paradigm by carrying over a procedural construct.
+
+Not everything is rosy, though. The main risk with Approach 2 is that the calling convention not explicited, which may lead to proliferation of calling conventions across user-deployed actors. In turn this may result in cognitive overhead and interoperability issues.
+
+An IPLD-based Interface Definition Language should eliminate that concern. 
+
+**Future: IPLD-based IDLs**
+
+IPLD Schemas are good for defining the structure of objects, but they do not help with expressing the interface of behaviours an object can offer (e.g. methods), inputs, and outputs.
+
+Our basic idea is to design an IPLD schema for IDLs; think of [gRPC IDLs](https://grpc.io/docs/what-is-grpc/core-concepts/), [SOAP WSDL](https://www.w3schools.com/xml/xml_wsdl.asp), etc. This schema would be standardised and versioned, and understood by the FVM SDK.
+
+This IDL would not have functions as such, but rather _labelled behaviours_. This is consistent with the actor model, where actors don't expose functions, but they are known to provide behaviours.
+
+Each behaviour would carry an IPLD Schema for the input type and the return value type, an error code enum, and potentially an identifier (which incidentally could be a method number!) to facilitate internal dispatch.
+
+Traditional "overloaded functions" can be represented as distinct labelled behaviours dispatching to the same identifier.
+
+The FVM SDK would offer utilities to pack/generate calls to another actor with their IDL.
+
+_Pseudocode:_
+
+```
+import "idls/other-actor.idl.ipld" as actor
+from sdk import calls
+
+// Res is typed automatically as Result<Type, Error>
+let res = calls.target(actor)
+               .call("behaviour")
+               .with({ Field2: Value, Field2: [Value1, Value2]})
+```
+
+**Test sandbox**
+
+The FVM project will need to ship a test actor sandbox for developers to use during their actor development process.
+
+**Foreign runtimes**
+
+Support for foreign runtimes, such as the EVM, will be introduced by emulating those runtimes over WASM by compiling the respective interpreter to WASM bytecode.
+
+We expect a performance penalty from doing this (the extent of it is to be determined), but this tradeoff is workable in exchange for the highest execution fidelity/parity.
```

Comment:
1 time perf penalty? or will we compile to a more performant format after the first?

Reply:
Performance penalty coming from the EVM => WASM interpretation. EVM => WASM transpilation would turn this into a one-time penalty, but transpliation remains an unresolved problem.